PSA Secure partition manager and services

orenc17 opened the pull request:
### Description

<!-- 
    Required
    Add here detailed changes summary, testing results, dependencies 
    Good example: https://os.mbed.com/docs/latest/reference/workflow.html (Pull request template)
-->
* Introduce PSA Secure partition manager (SPM) to mbed-os.
* Add SPM tests (for PSA targets).
* Add PSA SPM hal layer.
* Add PSA PRoT internal trusted storage Secure implementation.
* Integrate SPM into the boot proccess.
* PSA manifest data generator.
* Add PSA abstract targets to mbed-os targets.json.
* Add artifact delivery to the tools ( by @theotherjimmy ).

## Dependencies
#8667 
#8680 
#8730 

The actual changes of this PR are in commit ad7876e0620531cadaad7502f27cdd5f4d8c814f

### Pull request type

<!-- 
    Required
    Please add only one X to one of the following types. Do not fill multiple types (split the pull request otherwise).
    Please note this is not a GitHub task list, indenting the boxes or changing the format to add a '.' or '*' in front
    of them would change the meaning incorrectly. The only changes to be made are to add a description text under the
    description heading and to add a 'x' to the correct box.
-->
    [ ] Fix
    [ ] Refactor
    [ ] Target update
    [X] Functionality change
    [ ] Docs update
    [ ] Test update
    [ ] Breaking change

Commit:
PSA SPM

* Intorduce PSA-SPM to mbed-os
* Add SPM tests (for PSA targets)
* Add PSA PRoT internal storage Secure implementation
* Integrate SPM into the boot proccess
* PSA manifest data generator
* Introduce PSA targets skeleton to mbed-os
* Add artifact delivery to the tools

theotherjimmy commented:
Thanks for the credit!

theotherjimmy commented:
It looks like I forgot to update the python tests to include the new Config attribute. I'll add that later today.

theotherjimmy committed:
Include `deliver_into` in Config mock

Commit:
astyle changes

Commit:
Normalize license headers

- add SPDX license identifier
- add missing license headers
- update year in license headers

0xc0170 commented:
@ARMmbed/mbed-os-tools Please review tools changes

> dependencies

2 dependencies are still in progress:

#8667 - in progress
#8680 - merged
#8730 - depends on 8667

Commit:
Coverity bug fixes

Commit:
Fix the SPM HAL test

bulislaw reviewed (changes requested):
My comments can be found in the commit: https://github.com/ARMmbed/mbed-os/commit/ad7876e0620531cadaad7502f27cdd5f4d8c814f

0xc0170 commented on TESTS/mbed_hal/spm/fault_functions.h:
```diff
@@ -0,0 +1,48 @@
+/* mbed Microcontroller Library
+ * Copyright (c) 2017 ARM Limited
```
All new files should have 2018 year and using SPDX identifier

orenc17 replied:
fixed

0xc0170 commented on tools/settings.py:
```diff
@@ -27,6 +27,8 @@
 ##############################################################################
 BUILD_DIR = abspath(join(ROOT, "BUILD"))
 
+DELIVERY_DIR = abspath(join(ROOT, "DELIVERY"))
```
what does this do? Is it documented anywhere ( no comment here, cant find it in the commit msg neither)

orenc17 replied:
Added comment

0xc0170 commented on hal/spm_api.h:
```diff
@@ -0,0 +1,101 @@
+ *
+ * Wakeup the peer processor waiting on the mailbox driver event.
+ *
+ * @note The functions below should be implemented by target specific code.
```
quick question: these new docs were already reviewed by docs team or shall be done (within this pull request) ?

orenc17 replied:
@danny4478 and @AnotherButler should know the aswer

danny4478 replied:
@0xc0170 Which docs are you referring to?
- The code (c and h files) was not reviewed by docs team. Once the cod is ready for merge, the docs team will generate the Doxygen for them.
- Regarding docs, we have 3 documents (handbook) under review by the docs team.

0xc0170 replied:
Doxygen documentation (functionality docs) - they have to be reviewed by docs  team. 

Therefore if that is true, it would be helpful here provided what doxygen docs should be reviewed. I identified at least this header file

alzix replied:
Doxygen headers for review are:
[ ] - components/TARGET_PSA/spm/spm_client.h
[ ] - components/TARGET_PSA/spm/COMPONENT_SPE/spm_server.h
[ ] - components/TARGET_PSA/spm/psa_defs.h
[ ] - hal/spm_api.h

0xc0170 replied:
Thankss, this is helpful !

@melwee01 @AnotherButler Please review these

Commit:
Generate better magic numbers

Commit:
Remove internal RTX includes

Commit:
Add Version to SPM templates and generator

Commit:
Remove SPM negative tests

Commit:
Fix FVP_MPS2 & ARM_CM3DS_MPS2 components definition

Commit:
Handle manager documentation fix

Move function description to header files

Note: it is not in Doxygen format since it is an internal module and
        Doxygen should not be generated for it

Commit:
Remove dead code

0xc0170 commented:
We are waiting for the last dependency to land : #8730

0xc0170 reviewed (commented):
I've just noticed I did not publish this on Friday, since  I reviewed there were few comments so might be already addressed.

0xc0170 commented:
All dependencies were integrated, please rebase (I reviewed astyle failures, rebase should take care of them as well?)

Commit:
CR fixes

orenc17 commented:
Branch rebased and ready to go

NirSonnenschein commented:
@0xc0170 @bulislaw , fixes are in please take a look. I'll run this through CI in anticipation of review (previous version passed so hopefully this will not be an issue)

0xc0170 commented on TESTS/mbed_hal/spm/main.cpp:
```diff
@@ -0,0 +1,156 @@
+#endif
+
+#if (defined( __CC_ARM ) || defined(__ARMCC_VERSION) || defined( __ICCARM__ ))
+#error [NOT_SUPPORTED] this test is supported on GCC only
```
What limitations are here that only GCC ARM supported? ) ?

alzix replied:
mainly assembly code included in the test

0xc0170 commented on components/TARGET_PSA/spm/spm_init.h:
```diff
@@ -0,0 +1,39 @@
+
+#if defined(COMPONENT_SPE)
+
+void psa_spm_init(void);
```
why these do not have any documentation?

alzix replied:
done

0xc0170 commented on components/TARGET_PSA/spm/COMPONENT_SPE/spm_internal.h:
```diff
@@ -0,0 +1,237 @@
+    const uint32_t out_len
+);
+
+void channel_state_switch(uint8_t *current_state, uint8_t expected_state, uint8_t new_state);
```
these 2 functions do not have doxygen, the rest of the functions here have?

alzix replied:
done

danny4478 replied:
This is confusing.
This is an internal API.
AFAIU, it should not be Doxygen documented. Of course it should have normal code comments.
My bottom line is: This file comments should be be Doxygen style.

0xc0170 commented on components/TARGET_PSA/spm/COMPONENT_SPE/spm_main.c:
```diff
@@ -0,0 +1,34 @@
+
+#else
+
+#error "Non dual-chip platforms are not yet supported"
```
Is this error correct ( this is else for COMPONENT_SPM_MAILBOX) ?

orenc17 replied:
Yes, there should be another main function for secure-side on v8 systems.
as we are not yet supporting v8 with SPM, the compilation should fail

0xc0170 reviewed (commented):
Few questions

0xc0170 commented:
@ARMmbed/mbed-os-tools  Please review today

Commit:
spm_internal.h - add missing function docs

Commit:
spm_init.h add function docs

Commit:
Astyle fixes

Commit:
Doxygen fixes in tests

cmonr commented:
Making a note here. This appears to be undergoing email discussion. Will update once resolved.

bulislaw commented:
All the PRs need to be engineering ready (marked as "needs: CI") by the end of the day (Austin time). Otherwise it won't make 5.11 and will need to come in the next release (5.12 for features, 5.11.1 for fixes and new platforms).

bulislaw commented:
@SenRamakri Can you chase Jimmy for a review please.

theotherjimmy reviewed (approved):
Everything looks like I expect it to.

NirSonnenschein commented:
@bulislaw  tools have approved, what about you?

bulislaw commented on components/TARGET_PSA/spm/COMPONENT_SPM_MAILBOX/COMPONENT_NSPE/spm_client_proxy.c:
```diff
@@ -0,0 +1,229 @@
+
+uint32_t psa_version(uint32_t sid)
+{
+    osRtxSemaphore_t res_sem_storage = {0, 0, 0, 0, 0, 0, 0, 0};
```
We'll need to replace that with Mbed OS provided abstraction. I can accept it on condition that we will create a bug and fix it in 5.12. That's for all the places we use internal RTX types.

bulislaw commented on components/TARGET_PSA/spm/doc/INTRO.md:
```diff
@@ -0,0 +1,117 @@
+
+The secure partition manifest file describes the properties of the secure partitions. In this file:
+
+* **entry_point** is the function name of the partition's thread.
```
My comments from the initial commit weren't addressed in any way. Not a blocker, but it's weird to have example with some fields documented and some not. You should either document all of them or make the example simpler.

orenc17 replied:
fixed in #8873

bulislaw commented on hal/spm_api.h:
```diff
@@ -0,0 +1,103 @@
+#endif
+
+
+/** @addtogroup HAL-SPE
```
These docs won't be generated by doxygen unless you ad the COMPONENTS_SPE flag to: `doxyfile_options` and `doxygen_options.json`

orenc17 replied:
fixed in #8873

bulislaw reviewed (approved):
@alzix I conditionally approve the PR, but some things will need to be fixed:
* You'll need to follow with a small PR ASAP (for 5.11) to enable generating doxygen
* I really want usage of internal RTX types be contained to 1 file in Mbed, that will need to be updated after code freeze on master
* The example you provide should ideally not use not documented fields (you document some, but not all?)

cmonr commented:
CI started.

Commit:
Edit spm_client.h

Make minor copy edits, mostly for punctuation and style.

Commit:
Edit spm_server.h

Make minor edits, mostly for consistent capitalization and style.

Commit:
Edit psa_defs.h

Make minor edits, mostly for consistent abbreviations.

Commit:
Edit spm_api.h

Edit file, mostly for minor grammar and style changes.

AnotherButler reviewed (approved):
I've edited the requested files. I do have one query that doesn't need to prevent this from merging: What's a doorbell signal?

cmonr commented:
CI restarted.

Aborted due to out of space issue.

alzix commented on components/TARGET_PSA/spm/psa_defs.h:
```diff
@@ -44,7 +44,7 @@ extern "C" {
 
 #define PSA_FRAMEWORK_VERSION    (0x0009) /**< Version of the PSA Framework API. */
-#define PSA_VERSION_NONE         (0L)     /**< Identifier for an unimplemented RoT Service. */
+#define PSA_VERSION_NONE         (0L)     /**< Identifier for an unimplemented Rot of Trust (RoT) Service. */
```
it is Root of Trust :)
```suggestion
#define PSA_VERSION_NONE         (0L)     /**< Identifier for an unimplemented Root of Trust (RoT) Service. */
```

Commit:
Update components/TARGET_PSA/spm/psa_defs.h

Co-Authored-By: orenc17 <oren.cohen@arm.com>

Commit:
Fix targets.json

NirSonnenschein commented:
CI restarted

melwee01 commented on components/TARGET_PSA/spm/COMPONENT_SPE/handles_manager.h:
```diff
@@ -0,0 +1,160 @@
+
+/* -------------------------------- Handle Manager Module ---------------------------- */
+
+/* The Handle Manager Module manages handles.
```
Presumably. Do we need to actually state this?

melwee01 commented on components/TARGET_PSA/spm/COMPONENT_SPE/handles_manager.h:
```diff
@@ -0,0 +1,160 @@
+
+/* The Handle Manager Module manages handles.
+ *
+ * It basically generates and exposes a unique handle identifier [handle] per
```
Don't use "basically" or similar.

melwee01 replied:
(`handle`), (`handle_mem`), etc.

melwee01 commented on components/TARGET_PSA/spm/COMPONENT_SPE/handles_manager.h:
```diff
@@ -0,0 +1,160 @@
+ * - Remove a handle from the handle manager module [handle_destroy]
+ *
+ * Note:
+ * Handles generation is done exclusively.
```
Handle

melwee01 commented on components/TARGET_PSA/spm/COMPONENT_SPE/handles_manager.h:
```diff
@@ -0,0 +1,160 @@
+ *
+ * Note:
+ * Handles generation is done exclusively.
+ * Once we got a handle, removing a handle or getting its memory can be
```
~we~, use you instead

non-exclusively

melwee01 commented on components/TARGET_PSA/spm/COMPONENT_SPE/handles_manager.h:
```diff
@@ -0,0 +1,160 @@
+/*
+ * @brief create unique handle identifier
+ *
+ * This function generates a unique handle identifier, and "couples" it with the received handle memory.
```
Don't put things in quotation marks, please. Use ** for key words if necessary, but in this case, just remove the punctuation.

melwee01 commented on components/TARGET_PSA/spm/COMPONENT_SPE/handles_manager.h:
```diff
@@ -0,0 +1,160 @@
+ *
+ * @param[in] handle_mgr   A pointer to the handle manager object
+ * @param[in] handle_mem   A pointer to a pre-allocated handle memory to get a handle identifier for
+ * @param[in] friend_pid   The partition id which is allowed to get_mem() and destroy() in addition to the handle owner.
```
Put functions in ``

melwee01 commented on components/TARGET_PSA/spm/COMPONENT_SPE/handles_manager.h:
```diff
@@ -0,0 +1,160 @@
+ *       In case memory pool allocation fails, this function should not be called.
+ *       This function will panic on non vacant space use case.
+ *
+ * @param[in] handle_mgr   A pointer to the handle manager object
```
Inconsistent use of . at end

melwee01 commented on components/TARGET_PSA/spm/COMPONENT_SPE/handles_manager.h:
```diff
@@ -0,0 +1,160 @@
+ *
+ * This function retrieves the pointer associated with the input <handle>.
+ *
+ * @note  This function will panic in case caller not allowed to dereference the memory
```
Don't understand.

melwee01 commented on components/TARGET_PSA/spm/COMPONENT_SPE/handles_manager.h:
```diff
@@ -0,0 +1,160 @@
+ *         or handler does not correspond to a valid existing handle
+ *
+ * @param handle_mgr   A pointer to the handle manager object.
+ * @param handle       The handle for which we request the corresponding memory handle.
```
~we~ Use you instead

melwee01 commented on components/TARGET_PSA/spm/COMPONENT_SPE/spm_internal.h:
```diff
@@ -0,0 +1,251 @@
+    struct spm_partition *src_partition; /* Pointer to the Partition which connects to the Root of Trust Service.*/
+    spm_rot_service_t *dst_rot_service; /* Pointer to the connected Root of Trust Service.*/
+    void *rhandle; /* Reverse handle to be used for this channel.*/
+    void *msg_ptr; /* message data sent from user */
```
Message

melwee01 commented on components/TARGET_PSA/spm/COMPONENT_SPE/spm_internal.h:
```diff
@@ -0,0 +1,251 @@
+    spm_rot_service_t *dst_rot_service; /* Pointer to the connected Root of Trust Service.*/
+    void *rhandle; /* Reverse handle to be used for this channel.*/
+    void *msg_ptr; /* message data sent from user */
+    struct spm_ipc_channel *next; /* Next channel in the chain  */
```
inconsistent use of periods

melwee01 commented on components/TARGET_PSA/spm/COMPONENT_SPE/spm_internal.h:
```diff
@@ -0,0 +1,251 @@
+} spm_active_msg_t;
+
+/*
+ * Structure containing resources and attributes of a Secure Partition.
```
secure partition.

melwee01 commented on components/TARGET_PSA/spm/COMPONENT_SPE/spm_internal.h:
```diff
@@ -0,0 +1,251 @@
+    const uint32_t flags_interrupts; /* Mask of all the IRQs & doorbell which the partition supports.*/
+    spm_rot_service_t *rot_services; /* Array of the Partition's Root of Trust Services.*/
+    const uint32_t rot_services_count; /* Number of the Partition's Root of Trust Services.*/
+    const uint32_t *extern_sids; /* Array of Root of Trust Service IDs which the partition can connect to.*/
```
~which~ that

melwee01 commented on components/TARGET_PSA/spm/COMPONENT_SPE/spm_internal.h:
```diff
@@ -0,0 +1,251 @@
+// Platform dependent APIs
+
+/*
+ * Validates a memory block is accessable from a specific partition
```
Validates that or validates a memory block accessible from a specific partition (misspelling as well)

melwee01 commented on components/TARGET_PSA/spm/COMPONENT_SPE/spm_internal.h:
```diff
@@ -0,0 +1,251 @@
+ * Validates a memory block is accessable from a specific partition
+ *
+ * @param[in] ptr pointer to the beggining of the memory block.
+ * @param[in] size size of the memory block in bytes.
```
Capitalize beginning of description

melwee01 commented on components/TARGET_PSA/spm/COMPONENT_SPE/spm_internal.h:
```diff
@@ -0,0 +1,251 @@
+ * @param[in] ptr pointer to the beggining of the memory block.
+ * @param[in] size size of the memory block in bytes.
+ * @param[in] accessing_partition which partition is trying to access the memory.
+ * @return true if the entire memory block is accessable from given partition.
```
`true`

melwee01 commented on components/TARGET_PSA/spm/COMPONENT_SPE/spm_internal.h:
```diff
@@ -0,0 +1,251 @@
+/*
+ * Validates parameters sent from caller and queues a connect message on the correct ROT_SRV.
+ *
+ * @param[in] sid - desired RoT service ID
```
some h files have -, some have spaces, some have */: what's it supposed to be?

orenc17 replied:
not doxygen

melwee01 commented on components/TARGET_PSA/spm/COMPONENT_SPE/spm_server.h:
```diff
@@ -0,0 +1,172 @@
+#endif
+
+/** @addtogroup RoT-Service-API
+ *  The C interface for a Root of Trust Service in a partition.
```
root of trust service (has been shortened to RoT in other files)

melwee01 commented on components/TARGET_PSA/spm/COMPONENT_SPE/spm_server.h:
```diff
@@ -0,0 +1,172 @@
+/**
+ * Get the message that corresponds to a given signal.
+ *
+ * @param[in]  signum an asserted signal returned from psa_wait().
```
Capitalization here as well.

melwee01 commented on components/TARGET_PSA/spm/doc/INTRO.md:
```diff
@@ -0,0 +1,117 @@
+## The Secure Partition Manager
+
+The **Secure Partition Manager (SPM)** is a Platform Security Architecture (PSA) compliant software hypervisor that creates and manages independent Secure Partitions on Arm Cortex&reg;-M microcontrollers. It increases resilience against malware and protects secrets from leaking between different modules in the same application. The SPM complements other important security features, such as safe f ...
```
What's a hypervisor?

orenc17 replied:
the file will be removed

melwee01 commented on components/TARGET_PSA/spm/doc/INTRO.md:
```diff
@@ -0,0 +1,117 @@
+
+The **Secure Partition Manager (SPM)** is a Platform Security Architecture (PSA) compliant software hypervisor that creates and manages independent Secure Partitions on Arm Cortex&reg;-M microcontrollers. It increases resilience against malware and protects secrets from leaking between different modules in the same application. The SPM complements other important security features, such as safe f ...
+
+The SPM provides hardware-enforced partitions for individual code blocks by limiting access to memories and peripherals using the existing hardware security features of the Cortex&reg;-M microcontrollers. It isolates software in partitions, managing the execution of software within those partitions and providing Inter Process Communication (IPC) between the partitions. Correct use of SPM prevents ...
```
Cortex&reg;-M

Is this a render error?

melwee01 commented on components/TARGET_PSA/spm/doc/INTRO.md:
```diff
@@ -0,0 +1,117 @@
+
+The SPM and the secure partitions are located in the Secure Processing Environment (SPE), isolating them from the Non-Secure Processing Environment (NSPE), which contains the application firmware, OS kernel and libraries, and other nonsecure hardware resources.
+
+A secure partition is a container for one or more root of trust services, and a platform may have multiple secure partitions. Secure partitions provide the execution environment for security functionality.
```
one more root of trust services. A platform ...

melwee01 commented on components/TARGET_PSA/spm/doc/INTRO.md:
```diff
@@ -0,0 +1,117 @@
+
+If you are prototyping software or using platforms without SAU or MPU, you can choose to have no isolation between the SPE and NSPE (sometimes referred to as Level 0), though the PSA does not specify this. However, for production software, consider implementing one of the following levels of isolation:
+
+* **Level 1 - SPE isolation** In this level, the SPE is fully isolated from the nonsecure application firmware and hardware.
```
: instead of -

melwee01 commented on components/TARGET_PSA/spm/doc/README.md:
```diff
@@ -0,0 +1,23 @@
+# Mbed Secure Partition Manager (SPM)
+
+The Platform Security Architecture (PSA) firmware framework specifications contain a logic component called the Secure Partition Manager (SPM). PSA defines a Secure Processing Environment (SPE) for:
```
Mbed secure partition manager, unless it's a brand name. Is it?

orenc17 replied:
the file will be removed as well

melwee01 commented on components/TARGET_PSA/spm/doc/README.md:
```diff
@@ -0,0 +1,23 @@
+# Mbed Secure Partition Manager (SPM)
+
+The Platform Security Architecture (PSA) firmware framework specifications contain a logic component called the Secure Partition Manager (SPM). PSA defines a Secure Processing Environment (SPE) for:
+
```
Same as above comment for SPM and PSA. These are general terms, not brand names, right?

melwee01 commented on components/TARGET_PSA/spm/doc/README.md:
```diff
@@ -0,0 +1,23 @@
+The Platform Security Architecture (PSA) firmware framework specifications contain a logic component called the Secure Partition Manager (SPM). PSA defines a Secure Processing Environment (SPE) for:
+
+* Sensitive data, such as keys, credentials and firmware.
+* The code that manages it.
```
What's it?

melwee01 commented on components/TARGET_PSA/spm/doc/README.md:
```diff
@@ -0,0 +1,23 @@
+* The code that manages it.
+* Its trusted hardware resources.
+
+The PSA SPM interfaces decouple components, allowing reuse of components in other device platform and helps to reduce an integration effort.
```
nonparallel construction

melwee01 commented on components/TARGET_PSA/spm/doc/README.md:
```diff
@@ -0,0 +1,23 @@
+
+* Secures low cost IoT devices, where a full Trusted Execution Environment (TEE) would not be appropriate.
+* Protects sensitive assets (keys, credentials and firmware) by separating these from the application firmware and hardware.
+* Is architecture agnostic and can be implemented on different Arm Cortex&reg;-M architectures, offering variable level of protection, based on platform resources.
```
another character render error

0xc0170 commented:
Restarted CI , one device failed with serial error (in events test)

alzix commented:
@melwee01, thank you for review.
We will fix all your remarks in followup PR RC2
